lib/deep_cover/autoload_tracker.rb

```diff
@@ -0,0 +1,208 @@
+      @autoloads_by_basename[basename] << entry
+      interceptor_path
+    end
+
+    def entries_for_target(requested_path, absolute_path_found)
+      basename = basename_without_extension(requested_path)
+      autoloads = @autoloads_by_basename[basename] || []
+
+      if absolute_path_found
+        autoloads.select { |entry| entry_is_target?(entry, requested_path, absolute_path_found) }
+      elsif requested_path == File.absolute_path(requested_path)
+        []
+      elsif requested_path.start_with?('./', '../')
+        []
+      else
+        # We didn't find a path that goes through the $LOAD_PATH
+        # It's possible that RubyGems will actually add the $LOAD_PATH and require an actual file
+        # So we must make a best-guest for possible matches
+        requested_path_to_compare = without_extension(requested_path)
+        autoloads.select { |entry| requested_path_to_compare == without_extension(entry.target_path) }
+      end
+    end
+
+    def entry_is_target?(entry, requested_path, absolute_path_found)
+      return true if entry.target_path == requested_path
+      target_path_rb = with_rb_extension(entry.target_path)
+      return true if target_path_rb == requested_path
+      entry_absolute_path = DeepCover.custom_requirer.resolve_path(entry.target_path)
+      return true if entry_absolute_path == absolute_path_found
+      false
+    end
+
+    def basename_without_extension(path)
+      without_extension(File.basename(path))
+    end
+
+    def with_rb_extension(path)
+      path += '.rb' unless needs_extension?(path)
+      path
+    end
+
+    def without_extension(path)
+      path = path[0...-3] unless needs_extension?(path)
+      path
+    end
+
+    def needs_extension?(path)
+      !path.end_with?('.rb', '.so')
+    end
+
+    def autoload_interceptor_for(path)
+      existing_files = @interceptor_files_by_path[path] || []
+      reusable_file = existing_files.detect { |f| !$LOADED_FEATURES.include?(f.path) }
+      return reusable_file.path if reusable_file
+
+      new_file = Tempfile.new([File.basename(path), '.rb'])
+      # Need to store all the tempfiles so that they are not GCed, which would delete the files themselves.
+      # Keeping them by path allows us to reuse them.
+      @interceptor_files_by_path[path] ||= []
+      @interceptor_files_by_path[path] << new_file
+      new_file.write(<<-RUBY)
+# Intermediary file for ruby's autoload made by deep-cover
+require #{path.to_s.inspect}
+      RUBY
+      new_file.close
+
+      new_file.path
+    end
+
+    class << self
+      # Can be true, false, or a Symbol. The symbol will make every warning be displayed
+      attr_accessor :warned_for_frozen_module
+    end
+    self.warned_for_frozen_module = false
+
+    # Using frozen modules/classes is almost unheard of, but a warning makes things easier if someone does it
+    def warn_frozen_module(const)
+      return if self.class.warned_for_frozen_module == true
```

Comment:
More nitpicking: remove `== true`

Reply:
The value can also be a symbol. `:every` or `:all`... or anything, the Symbol makes all warnings appear.

Reply:
Ah, that "feature" is not tested (and seems way overkill! this is a warning for something that "is almost unheard of"). YAGNI + KISS = kill

Reply:
It was useful when I wanted to debug it, cuz we do have tests here and there that create frozen modules... Its not something meant for users to be honest. I can remove it if you prefer.

Reply:
Yeah, I'd kill it. I think it's a much better idea to have a `before { warned_for_frozen_module = false }` in the test suite.

Reply:
Done